llmfoundry/data/finetuning/collator.py

```diff
@@ -13,6 +13,168 @@
+        raise ValueError(
+            f'target_responses must be either "last" or "all" but {target_responses=}'
+        )
+
+    if target_prompts.startswith('length>='):
+        cutoff = target_prompts[8:]
+        if not cutoff.isdigit():
+            raise ValueError(
+                f'target_prompts starts with "length>=" but the rest of the string is not digits ({target_prompts=}). ' +\
+                'To use this configuration option, set target_prompts "length>=XX" where "XX" is a positive integer indicating ' +\
+                'the length cutoff. Prompts of at least XX tokens in length will be treated as targets.'
+            )
+        cutoff = int(cutoff)
+        if cutoff <= 0:
+            raise ValueError(
+                f'You are trying to set the target_prompts length cutoff to a negative number {cutoff=}. This is not allowed.'
+            )
+    elif target_prompts not in {'all', 'none'}:
+        raise ValueError(
+            f'target_prompts must either be "all", "none" or "length>=XX" where "XX" is a positive integer, but {target_prompts=}'
+        )
+
+
+###### Functions to implement target_prompts and target_responses choices #####
+def _sequence_to_labels_all(sequence: list[int],
+                            is_last_turn: bool,
+                            cutoff: Optional[int] = None) -> list[int]:
+    del is_last_turn, cutoff  # unused
+    return sequence
+
+
+def _sequence_to_labels_none(sequence: list[int],
+                             is_last_turn: bool,
+                             cutoff: Optional[int] = None) -> list[int]:
+    del is_last_turn, cutoff  # unused
+    return [_HF_IGNORE_INDEX] * len(sequence)
+
+
+def _sequence_to_labels_last(sequence: list[int],
+                             is_last_turn: bool,
+                             cutoff: Optional[int] = None) -> list[int]:
+    del cutoff  # unused
+    if is_last_turn:
+        return sequence
+    else:
+        return [_HF_IGNORE_INDEX] * len(sequence)
+
+
+def _sequence_to_labels_cutoff(sequence: list[int],
+                               is_last_turn: bool,
+                               cutoff: Optional[int] = None) -> list[int]:
+    del is_last_turn  # unused
+    if cutoff is None:
+        raise ValueError('input ``cutoff`` must be provided')
+    if len(sequence) >= cutoff:
+        return sequence
+    else:
+        return [_HF_IGNORE_INDEX] * len(sequence)
+
+
+_TARGET_POLICY_LOOKUP = {
+    'all': _sequence_to_labels_all,
+    'none': _sequence_to_labels_none,
+    'last': _sequence_to_labels_last,
+    'length': _sequence_to_labels_cutoff,
+}
+
+
+def stitch_turns_decoder_only(
+        example_turns: list[dict[str, list[int]]],
+        target_prompts: str,
+        target_responses: str,
+        eos_token_id: Optional[int] = None,
+        validate: bool = False) -> tuple[list[int], list[int]]:
+    target_prompts = target_prompts.lower()
+    target_responses = target_responses.lower()
+
+    if validate:
```

Comment:
why is validation optional?

Reply:
Just to add flexibility I guess. There are other validation checks that should catch invalid settings in the dataloader and in the convert-to-streaming script. So, if we did validation every time this function was run, it'd be redundant with those checks and waste time. But I figured there might be times this function would want to validate on its own, so I added it as an option. And I figure this will be useful if we eventually allow each sample to have its own target prompts/responses settings.